### PR TITLE
ENH: Use contextlib to make context managers

### PR DESCRIPTION
**Pull request by kohr-h**
Using `contextlib.contextmanager`, we can scrap the boilerplate stuff (`__enter__` and `__exit__`, `__init__`) and write context managers in a much more straightforward way with plain functions and `try...finally`. A side benefit is that their usage looks more like a function call than a class creation.

All of our current custom context managers have been ported.

### CODE REVIEW

File: odl/util/testutils.py [hunk 1]
```diff
@@ -459,80 +460,78 @@ def noise_elements(space, n=1):
...
+        *** FAILED 1 TEST CASE(S) ***
     """
 
-    def __init__(self, test_name, err_msg=None, logger=print):
-        self.num_failed = 0
-        self.test_name = test_name
-        self.err_msg = err_msg
-        self.fail_strings = []
-        self.log = logger
+    class _Counter(object):
```
Comment (adler-j) [inline odl/util/testutils.py]: _FailCounter perhaps?

File: odl/util/testutils.py [hunk 2]
```diff
@@ -459,80 +460,78 @@ def noise_elements(space, n=1):
...
+    tstart = None
+    try:
+        tstart = time()
+        yield
 
-    def __exit__(self, type, value, traceback):
-        time_str = '{:.3f}'.format(time() - self.tstart)
-        print('{:>30s} : {:>10s} '.format(self.name, time_str))
+    finally:
+        if tstart is not None:
```
Comment (adler-j) [inline odl/util/testutils.py]: Could this ever not happen?

File: odl/util/utility.py [hunk 3]
```diff
@@ -1533,41 +1534,37 @@ def pkg_supports(feature, pkg_version, pkg_feat_dict):
...
+    >>> rand_int == same_rand_int
+    True
+    """
+    do_seed = seed is not None
+    try:
+        if do_seed:
+            orig_rng_state = np.random.get_state()
+            np.random.seed(seed)
+        yield
+
```
Comment (adler-j) [inline odl/util/utility.py]: Remove blankline here

File: odl/util/testutils.py (hunk 2 above)
Comment (kohr-h) [reply odl/util/testutils.py]: Yeah I'm very cautious here. If `time()` fails without raising an exception, then we're in serious trouble anyway.

File: odl/util/testutils.py (hunk 1 above)
Comment (kohr-h) [reply odl/util/testutils.py]: Sure!

File: odl/util/testutils.py (hunk 2 above)
Comment (kohr-h) [reply odl/util/testutils.py]: Removed that precaution, `time` says it returns a float.

**Pull request merged**